Implement gather primitive for ProcessGroupNCCL

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #71623
* #70029
* __->__ #66745

This PR implement NCCL gather and add gather to ProcessGroupNCCL using nccl send/recv api.

NCCL doesn’t directly provide primitives for gather, so we need to be implemented on top of NCCL’s send/recv API.
1. In ProcessGroupNCCL.cpp, the outputTensors are first flattened, then inputTensors and outputFlattened are passed by the collective class to gather() function in nccl.cpp.
1. In nccl.cpp, gather is implemented using ncclSend/ncclRecv: all the ranks send inputTensor to the root rank, and the root rank uses a for loop to receive these inputTensors.

Differential Revision: [D29616361](https://our.internmc.facebook.com/intern/diff/D29616361/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29616361/)!

Commit:
Implement gather primitive for ProcessGroupNCCL

This PR implement NCCL gather and add gather to ProcessGroupNCCL using nccl send/recv api.

NCCL doesn’t directly provide primitives for gather, so we need to be implemented on top of NCCL’s send/recv API.
1. In ProcessGroupNCCL.cpp, the outputTensors are first flattened, then inputTensors and outputFlattened are passed by the collective class to gather() function in nccl.cpp.
1. In nccl.cpp, gather is implemented using ncclSend/ncclRecv: all the ranks send inputTensor to the root rank, and the root rank uses a for loop to receive these inputTensors.

Differential Revision: [D29616361](https://our.internmc.facebook.com/intern/diff/D29616361/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29616361/)!

[ghstack-poisoned]

Commit:
Update on "Implement gather primitive for ProcessGroupNCCL"


This PR implement NCCL gather and add gather to ProcessGroupNCCL using nccl send/recv api.

NCCL doesn’t directly provide primitives for gather, so we need to be implemented on top of NCCL’s send/recv API.
1. In ProcessGroupNCCL.cpp, the outputTensors are first flattened, then inputTensors and outputFlattened are passed by the collective class to gather() function in nccl.cpp.
1. In nccl.cpp, gather is implemented using ncclSend/ncclRecv: all the ranks send inputTensor to the root rank, and the root rank uses a for loop to receive these inputTensors.

Differential Revision: [D29616361](https://our.internmc.facebook.com/intern/diff/D29616361/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29616361/)!

[ghstack-poisoned]

wanchaol commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -464,6 +464,215 @@ def allgather_base(output_t, input_t):
             allgather_base(output_t, tensor)
 
+    @requires_nccl()
+    def test_gather_ops(self):
```
I have tried merging these two actually, one problem is that ProcessGroupNCCL and ProcessGroupGloo behavior is different in terms of result for example (see https://github.com/pytorch/pytorch/issues/68478), I will work on making them consistent as a follow up diff, then merging the tests together

Commit:
Update on "Implement gather primitive for ProcessGroupNCCL"


This PR implement NCCL gather and add gather to ProcessGroupNCCL using nccl send/recv api.

NCCL doesn’t directly provide primitives for gather, so we need to be implemented on top of NCCL’s send/recv API.
1. In ProcessGroupNCCL.cpp, the outputTensors are first flattened, then inputTensors and outputFlattened are passed by the collective class to gather() function in nccl.cpp.
1. In nccl.cpp, gather is implemented using ncclSend/ncclRecv: all the ranks send inputTensor to the root rank, and the root rank uses a for loop to receive these inputTensors.

Differential Revision: [D29616361](https://our.internmc.facebook.com/intern/diff/D29616361/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29616361/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/cuda/nccl.cpp:
```diff
@@ -811,6 +811,47 @@ void recv(
+  {
+    for (int r = 0; r < numranks; r++)
+    {
+      auto* recvbuff =  reinterpret_cast<char*>(outputs[r].data_ptr());
```
nit: outputs[r] needs to create a new Tensor object (not necessarily a copy) and might have some minor overhead.. Was wondering if this can just be done using output.data_ptr().

pritamdamania87 replied:
The `[]` operator has a bunch of overhead and creates a new Tensor object: https://codebrowser.bddppq.com/pytorch/pytorch/aten/src/ATen/TensorOperators.h.html#_ZNK2at6TensorixEN3c106ScalarE. It might be slightly faster to just use `.data_ptr()` directly.

pritamdamania87 replied:
Although, I feel this is probably fine for now since directly manipulating `data_ptr()` would involve getting strides and dtype etc and some pointer arithmetic.

pritamdamania87 commented on torch/csrc/cuda/nccl.cpp:
```diff
@@ -811,6 +811,47 @@ void recv(
+  const auto* sendbuff = reinterpret_cast<char*>(inputs.data_ptr());
+
+  NCCL_CHECK(ncclGroupStart());
+  NCCL_CHECK(ncclSend(sendbuff, count, type, root, comm, stream));
```
The root rank shouldn't have to send to self?

Commit:
Update on "Implement gather primitive for ProcessGroupNCCL"


This PR implement NCCL gather and add gather to ProcessGroupNCCL using nccl send/recv api.

NCCL doesn’t directly provide primitives for gather, so we need to be implemented on top of NCCL’s send/recv API.
1. In ProcessGroupNCCL.cpp, the outputTensors are first flattened, then inputTensors and outputFlattened are passed by the collective class to gather() function in nccl.cpp.
1. In nccl.cpp, gather is implemented using ncclSend/ncclRecv: all the ranks send inputTensor to the root rank, and the root rank uses a for loop to receive these inputTensors.

Differential Revision: [D29616361](https://our.internmc.facebook.com/intern/diff/D29616361/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29616361/)!

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2142,10 +2142,95 @@ void ProcessGroupNCCL::groupEnd() {
+      rank_,                    // rank
+      "gather",             // colName
+      tensor.numel(),           // inSize
+      tensor.numel() *          // outSize
+        this->getSize(),        // dType
+      tensor.scalar_type(),
```
Nit: looks like comment and code are not aligned? `// outSize` should be in L2165, `// dType` should be in L 2166?

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2142,10 +2142,95 @@ void ProcessGroupNCCL::groupEnd() {
+      invalidArgument("requires empty output on non-root");
+    }
+
+    for(int i = 0; i < inputTensors.size(); ++ i) {
```
Nit: i++? (or ++i?)

Commit:
Update on "Implement gather primitive for ProcessGroupNCCL"


This PR implement NCCL gather and add gather to ProcessGroupNCCL using nccl send/recv api.

NCCL doesn’t directly provide primitives for gather, so we need to be implemented on top of NCCL’s send/recv API.
1. In ProcessGroupNCCL.cpp, the outputTensors are first flattened, then inputTensors and outputFlattened are passed by the collective class to gather() function in nccl.cpp.
1. In nccl.cpp, gather is implemented using ncclSend/ncclRecv: all the ranks send inputTensor to the root rank, and the root rank uses a for loop to receive these inputTensors.

Differential Revision: [D29616361](https://our.internmc.facebook.com/intern/diff/D29616361/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29616361/)!

[ghstack-poisoned]

Commit:
Update on "Implement gather primitive for ProcessGroupNCCL"


This PR implement NCCL gather and add gather to ProcessGroupNCCL using nccl send/recv api.

NCCL doesn’t directly provide primitives for gather, so we need to be implemented on top of NCCL’s send/recv API.
1. In ProcessGroupNCCL.cpp, the outputTensors are first flattened, then inputTensors and outputFlattened are passed by the collective class to gather() function in nccl.cpp.
1. In nccl.cpp, gather is implemented using ncclSend/ncclRecv: all the ranks send inputTensor to the root rank, and the root rank uses a for loop to receive these inputTensors.

Differential Revision: [D29616361](https://our.internmc.facebook.com/intern/diff/D29616361/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29616361/)!

[ghstack-poisoned]

Commit:
Update on "Implement gather primitive for ProcessGroupNCCL"


This PR implement NCCL gather and add gather to ProcessGroupNCCL using nccl send/recv api.

NCCL doesn’t directly provide primitives for gather, so we need to be implemented on top of NCCL’s send/recv API.
1. In ProcessGroupNCCL.cpp, the outputTensors are first flattened, then inputTensors and outputFlattened are passed by the collective class to gather() function in nccl.cpp.
1. In nccl.cpp, gather is implemented using ncclSend/ncclRecv: all the ranks send inputTensor to the root rank, and the root rank uses a for loop to receive these inputTensors.

Differential Revision: [D29616361](https://our.internmc.facebook.com/intern/diff/D29616361/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29616361/)!

[ghstack-poisoned]

Commit:
Update on "Implement gather primitive for ProcessGroupNCCL"


This PR implement NCCL gather and add gather to ProcessGroupNCCL using nccl send/recv api.

NCCL doesn’t directly provide primitives for gather, so we need to be implemented on top of NCCL’s send/recv API.
1. In ProcessGroupNCCL.cpp, the outputTensors are first flattened, then inputTensors and outputFlattened are passed by the collective class to gather() function in nccl.cpp.
1. In nccl.cpp, gather is implemented using ncclSend/ncclRecv: all the ranks send inputTensor to the root rank, and the root rank uses a for loop to receive these inputTensors.

Differential Revision: [D29616361](https://our.internmc.facebook.com/intern/diff/D29616361/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29616361/)!

[ghstack-poisoned]

Commit:
Update on "Implement gather primitive for ProcessGroupNCCL"


This PR implement NCCL gather and add gather to ProcessGroupNCCL using nccl send/recv api.

NCCL doesn’t directly provide primitives for gather, so we need to be implemented on top of NCCL’s send/recv API.
1. In ProcessGroupNCCL.cpp, the outputTensors are first flattened, then inputTensors and outputFlattened are passed by the collective class to gather() function in nccl.cpp.
1. In nccl.cpp, gather is implemented using ncclSend/ncclRecv: all the ranks send inputTensor to the root rank, and the root rank uses a for loop to receive these inputTensors.

Differential Revision: [D29616361](https://our.internmc.facebook.com/intern/diff/D29616361/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29616361/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2151,7 +2278,7 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::gather(
   assertSingleElementInput(invalidArgument, inputTensors);
 
-  std::vector<at::Tensor> outputFlattened;
+  // std::vector<at::Tensor> outputFlattened;
```
nit: remove completely?

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1438,6 +1438,116 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
+                  : c10::nullopt);
+
+  // Store references to outputs to be used by WorkNCCL::result and operator<<.
+  if (opType == OpType::GATHER) {
```
Why do we have this conditional here?

wanchaol replied:
This is because for Gather and Scatter, we are actually defining `outputs` differently, for `gather`, `outputs` is `tensor_list`, while for `scatter`, `outputs` is defined as `tensor`

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1438,6 +1438,116 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
+template <typename Fn, typename PreProcess, typename PostProcess>
+c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::oneAndAll(
+    std::vector<at::Tensor>& tensor,
+    std::vector<std::vector<at::Tensor>>& tensor_list,
```
Do we really need a new method like this that has a lot of similarities to `collective`? What if we just make `collective` take a `vector<vector<Tensor>>` instead of just `vector<Tensor>`? Then modify current code that passes in a `vector<Tensor>` to pass in a `vector<vector<Tensor>>` of size 1?

wanchaol replied:
Yeah I tried to consolidate all into `collective` before introducing this oneAndAll, failed mainly because for `gather` and `scatter`, the inputs and outputs are different (they are swapped), so need someway to distinguish it (i.e. we only can record input stream on one rank for `scatter`, not all ranks as its only available on rootRank, same for `gather`), so it was pretty hard to generalize. Let me try again and see if this could work

wanchaol replied:
I actually found a way to work around this issue, by not introducing `oneAndAll`, and still reuse `collective` to do the job, the trick is to ignore the inputs/outputs specified in `fn` and use function capture to deal with `tensor` and `tensor_list` directly. Updated the PRs to reflect the changes. Let me know if this works :)

Commit:
Update on "Implement gather primitive for ProcessGroupNCCL"


This PR implement NCCL gather and add gather to ProcessGroupNCCL using nccl send/recv api.

NCCL doesn’t directly provide primitives for gather, so we need to be implemented on top of NCCL’s send/recv API.
1. In ProcessGroupNCCL.cpp, the outputTensors are first flattened, then inputTensors and outputFlattened are passed by the collective class to gather() function in nccl.cpp.
1. In nccl.cpp, gather is implemented using ncclSend/ncclRecv: all the ranks send inputTensor to the root rank, and the root rank uses a for loop to receive these inputTensors.

Differential Revision: [D29616361](https://our.internmc.facebook.com/intern/diff/D29616361/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29616361/)!

[ghstack-poisoned]

Commit:
Update on "Implement gather primitive for ProcessGroupNCCL"


This PR implement NCCL gather and add gather to ProcessGroupNCCL using nccl send/recv api.

NCCL doesn’t directly provide primitives for gather, so we need to be implemented on top of NCCL’s send/recv API.
1. In ProcessGroupNCCL.cpp, the outputTensors are first flattened, then inputTensors and outputFlattened are passed by the collective class to gather() function in nccl.cpp.
1. In nccl.cpp, gather is implemented using ncclSend/ncclRecv: all the ranks send inputTensor to the root rank, and the root rank uses a for loop to receive these inputTensors.

Differential Revision: [D29616361](https://our.internmc.facebook.com/intern/diff/D29616361/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29616361/)!

[ghstack-poisoned]

Commit:
Update on "Implement gather primitive for ProcessGroupNCCL"


This PR implement NCCL gather and add gather to ProcessGroupNCCL using nccl send/recv api.

NCCL doesn’t directly provide primitives for gather, so we need to be implemented on top of NCCL’s send/recv API.
1. In ProcessGroupNCCL.cpp, the outputTensors are first flattened, then inputTensors and outputFlattened are passed by the collective class to gather() function in nccl.cpp.
1. In nccl.cpp, gather is implemented using ncclSend/ncclRecv: all the ranks send inputTensor to the root rank, and the root rank uses a for loop to receive these inputTensors.

Differential Revision: [D29616361](https://our.internmc.facebook.com/intern/diff/D29616361/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29616361/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2140,10 +2140,74 @@ void ProcessGroupNCCL::groupEnd() {
+    if (outputTensors.size() != 0) {
+      invalidArgument("requires empty output on non-root");
+    }
+    outputTensors.emplace_back();
```
nit: probably not a good idea to modify the variable passed in. Maybe just create an empty vector<at:Tensor> and pass that in instead?

Commit:
Update on "Implement gather primitive for ProcessGroupNCCL"


This PR implement NCCL gather and add gather to ProcessGroupNCCL using nccl send/recv api.

NCCL doesn’t directly provide primitives for gather, so we need to be implemented on top of NCCL’s send/recv API.
1. In ProcessGroupNCCL.cpp, the outputTensors are first flattened, then inputTensors and outputFlattened are passed by the collective class to gather() function in nccl.cpp.
1. In nccl.cpp, gather is implemented using ncclSend/ncclRecv: all the ranks send inputTensor to the root rank, and the root rank uses a for loop to receive these inputTensors.

Differential Revision: [D29616361](https://our.internmc.facebook.com/intern/diff/D29616361/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29616361/)!

[ghstack-poisoned]

Commit:
Update on "Implement gather primitive for ProcessGroupNCCL"


This PR implement NCCL gather and add gather to ProcessGroupNCCL using nccl send/recv api.

NCCL doesn’t directly provide primitives for gather, so we need to be implemented on top of NCCL’s send/recv API.
1. In ProcessGroupNCCL.cpp, the outputTensors are first flattened, then inputTensors and outputFlattened are passed by the collective class to gather() function in nccl.cpp.
1. In nccl.cpp, gather is implemented using ncclSend/ncclRecv: all the ranks send inputTensor to the root rank, and the root rank uses a for loop to receive these inputTensors.

Differential Revision: [D29616361](https://our.internmc.facebook.com/intern/diff/D29616361/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29616361/)!

[ghstack-poisoned]

Commit:
Update on "Implement gather primitive for ProcessGroupNCCL"


This PR implement NCCL gather and add gather to ProcessGroupNCCL using nccl send/recv api.

NCCL doesn’t directly provide primitives for gather, so we need to be implemented on top of NCCL’s send/recv API.
1. In ProcessGroupNCCL.cpp, the outputTensors are first flattened, then inputTensors and outputFlattened are passed by the collective class to gather() function in nccl.cpp.
1. In nccl.cpp, gather is implemented using ncclSend/ncclRecv: all the ranks send inputTensor to the root rank, and the root rank uses a for loop to receive these inputTensors.

Differential Revision: [D29616361](https://our.internmc.facebook.com/intern/diff/D29616361/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29616361/)!

[ghstack-poisoned]
